src/mono/mono/mini/aot-compiler.c

```diff
@@ -14355,6 +14392,217 @@ add_interp_in_wrapper_for_sig (MonoAotCompile *acfg, MonoMethodSignature *sig)
+	DIRECT_PINVOKE_SKIP,
+	DIRECT_PINVOKE_PARSED,
+	DIRECT_PINVOKE_ERR,
+} DirectPInvokeParsedResult;
+
+//---------------------------------------------------------------------------------------
+//
+// parsed_direct_pinvoke parses the direct pinvoke for the module_name and entrypoint_name
+// It presumes that is_direct_pinvoke_parsable (direct_pinvoke) returned true.
+//
+// Parsing:
+// The direct pinvoke is stripped of any leading or trailing whitespace.
+// The remaining is assumed to be of the form MODULE or MODULE!ENTRYPOINT.
+// The passed in pointers are then reassigned to the corresponding strings.
+//
+// Arguments:
+//  * acfg - the MonoAotCompiler instance
+//  * dpi (direct pinvoke) - the string corresponding to the direct pinvoke to parse
+//  ** module_name_ptr - the pointer to the module name string
+//  ** entrypoint_name_ptr - the pointer to the entrypoint name string
+//
+// Return Value:
+//  enum pertaining to whether or not the direct pinvoke entry should be skipped, was
+//  successfully split into the respective module_name and entrypoint_name, or if
+//  an error has occurred.
+//
+
+static DirectPInvokeParsedResult
+parsed_direct_pinvoke (MonoAotCompile *acfg, char *dpi, char **module_name_ptr, char **entrypoint_name_ptr)
+{
+	DirectPInvokeParsedResult res = DIRECT_PINVOKE_PARSED;
+	char *direct_pinvoke = g_strdup (dpi);
+	if (!direct_pinvoke) {
+		aot_printerrf (acfg, "Failed to strdup the direct pinvoke '%s'.\n", dpi);
+		res = DIRECT_PINVOKE_ERR;
+		goto early_exit;
+	}
+
+	if (!g_strstrip (direct_pinvoke)) {
+		res = DIRECT_PINVOKE_SKIP;
+		goto cleanup;
+	}
+
+	char **direct_pinvoke_split = g_strsplit (direct_pinvoke, "!", 2);
+	if (!direct_pinvoke_split) {
+		aot_printerrf (acfg, "Failed to split the provided 'direct_pinvoke' AOT option '%s' with delimiter '!'\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+		goto cleanup;
+	}
+
+	*module_name_ptr = g_strdup (direct_pinvoke_split[0]);
+	if (!*module_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the module name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+	}
+	*entrypoint_name_ptr = g_strdup (direct_pinvoke_split[1]);
+	if (direct_pinvoke_split[1] && !*entrypoint_name_ptr) {
+		aot_printerrf (acfg, "Failed to strdup the entrypoint name portion of direct pinvoke '%s'.\n", direct_pinvoke);
+		res = DIRECT_PINVOKE_ERR;
+	}
+	g_strfreev (direct_pinvoke_split);
+	direct_pinvoke_split = NULL;
+
+cleanup:
+	g_free (direct_pinvoke);
+	direct_pinvoke = NULL;
+
+early_exit:
+	return res;
+}
+
+//---------------------------------------------------------------------------------------
+//
+// add_direct_pinvoke adds the module and entrypoint of a specified direct pinvoke to the
+// MonoAotCompile instance's HashTable of module/entrypoint entries.
+// It is presumed that module_name_ptr and entrypoint_name_ptr point to valid strings.
+// It transfers ownership of the module_name and entrypoint_name strings to the HashTable
```

Comment:
Maybe comment that function takes ownership of strings in both success and failure cases and set them to NULL.

Reply:
I tried to clarify the comments to what I think is happening
```
// It transfers ownership of the module_name and entrypoint_name strings data to the HashTable
// This function takes ownership of the module and entrypoint strings pointers regardless of 
// success or failure (in which it frees the memory) and sets them to NULL.
```